duckduckgo/iOS
# Add Bookmark Folders

Task/Issue URL: https://app.asana.com/0/72649045549333/1143141356446970/f
Tech Design URL: https://app.asana.com/0/481882893211075/1200283968383818/f, https://app.asana.com/0/481882893211075/1201023911968241/f
CC:

**Description**:
Adds bookmark folders, including migrating bookmark and favourite storage from UserDefaults to CoreData.

**Steps to test this PR**:
Bookmarks/favorites appear in many parts of the app, so the test plan itself below should be scrutinised for anything missing. I've broken it down by area of the app to hopefully make that easier

**Test Migration**
1. To start, before you run this build, make sure you have favorites and bookmarks on a previous version of the app, and screenshot them, so you can compare their existence/position both pre and post migration
2. Install this PR
3. Open the app and go to the bookmark screen and check all favourites and bookmarks are in the right place (their ordering should be preserved)
 ---
1. Start with a previous build of the app again, with favs and bookmarks 
2. Install this PR 
3. Before you go to the bookmarks screen, add some favorites and bookmarks
4. Then go to the bookmarks screen, and check previous favs and bookmarks are in the right positions, and any new ones were correctly added to the end.
 ---
1. Delete any existing DDG app
2. Install this PR
3. Go to the bookmarks screen, ensure there is nothing there
4. Add a folder/bookmark/whatever, check it works correctly

**The browsing menu**
1. Go to a website that isn't a fav or bookmark
2. Check that the browsing menu shows the options to add it as a bookmark and a favorite
3. Add it as a bookmark from the browsing menu
4. Add it as a favourite from the browsing menu
5. Now when opening the browsing menu it should show the options to edit the bookmark, and remove the favorite
6. Edit the bookmark from the browsing menu, make sure the edit screen is the correct one for the bookmark and not the favourite.
7. Make some change (e.g. to the title) and save
8. Go to the bookmark screen to check the change saved correctly 
9. Go back to the browsing menu and remove the favourite
10. Open the browsing menu and check that the remove favourite option has changed to add favourite, and the bookmark option still says edit bookmark.
 ---
1. Create a bookmark from the browsing menu again
2. When the toast appears at the bottom of the screen to say the bookmark is created, press the edit button
3. Edit the bookmark and check it works as above
---
1. Create a favorite from the browsing menu 
2. When the toast appears, press undo. Check that the favorite doesn't exist
3. Do the reverse (remove a favourite from the browsing menu and press undo)

**The bookmarks screen**
1. Check that bookmarks and favourites show the correct favicons
2. Press on both a favorite and a bookmark, check they open the sites correctly
3. Swipe right on both a bookmark and a favorite and share them
4. Swipe left on a bookmark, a favorite and a folder and delete them. Check an empty folder doesn't show the confirmation prompt
5. Attempt to delete a folder with contents, it should show a confirmation prompt, check both options work as expected.
6. Press on a folder and check it shows the correct subfolder screen. 
---
1. Press the edit button, the cells should change to show the edit view
2. Rearrange some bookmarks and favorites 
3. Rearrange a bookmark into a favourite and vice versa
4. Attempting to do likewise with a folder, it shouldn't let you convert it to a folder.
5. Also test arranging bookmarks/folders from a sub folder.
---
1. When there are no bookmarks or favorites, check it shows the correct empty state
2. When a subfolder has no items, check it shows the correct empty state (different from the top level empty state)
---
1. Edit a bookmark, select a new parent folder
2. Save it, and check all views etc updated correctly 
---
1. Edit a bookmark, but press add a new folder
2. After creating a new folder, it should be selected automatically yon the edit bookmark screen
---
1. Edit a favorite, checks it doesn't show the select folder view
2. Make some changes and test saving it works 
---
1. Edit a favorite or bookmark, but don't save the changes. Check that no changes are saved
---
1. Edit a folder
2. Check that in the select folder view itself or any of its children don't appear (i.e. you shouldn't be able to add a folder to itself)

**Favorites in various places**
1. On iPad, check that the favourites overlay appears and its contents is still correct when using the url bar
---
1. Check that favorutes appear on the home screen
2. Long press and check that both editing and deleting favorites works from the home screen.
3. Whist on the Home Screen, open the bookmarks view and delete or edit a favorite, check this change is reflected in the Home Screen view

**Search**
1. From the bookmarks screen, scroll down to reveal the search bar
2. Try search terms that should show a) a bookmark b) a favorite, c) both, d) neither, and check the results are correct
3. Tap on a result to check that you can open a bookmark from the search results
4. Try to swipe to delete a search result, it shouldn't work
5. Try to swipe to share a search result, it _should_ work
Note: Folders should not appear in search results 
---
1. Type in URL bar search terms as above, check they are correct and the sites open correctly.


**Widgets**
1. When you have favorites already, add either the medium or large widget (i.e. one with favorites)
2. Check that the favorites appear there as expected (in the right order, with the right favicons etc)
3. Tap on a favourite and check it opens the site as expected
---
1. When you have no favorites and have not already been through favorites onboarding, add the widget
2. It should show the onboarding prompt, press on it
3. The app should show the favourites onboarding 
4. Once you've added a favorite, check the widget shows the new favorite
5. Then delete the fav, and check the change is propagated to the widget
---
1. Convert a bookmark to a favorites, or vice versa, and check the change is propagated to the widget

**Other**
1. Press the bookmark all tabs button from tabs screen, check that it does that.

**Pixels**
Test the following pixels fire appropriately
- case bookmarksFolderCreated = "m_bookmarks_folder_created"
- case bookmarkCreatedAtTopLevel = "m_bookmark_created_at_top_level"
- case bookmarkCreatedInSubfolder = "m_bookmark_created_in_subfolder"
- case bookmarkEditedAtTopLevel = "m_bookmark_edited_at_top_level"
- case bookmarkEditedInSubfolder = "m_bookmark_edited_in_subfolder"



**Copy Testing**:


**Orientation Testing**:


**Device Testing**:


**OS Testing**:


**Theme Testing**:


---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)
**When ready for review, remember to post the PR in MM**
